kernel/work.c

```diff
@@ -0,0 +1,966 @@
+	__ASSERT_NO_MSG(sync != NULL);
+	__ASSERT_NO_MSG(!atomic_test_bit(&work->flags, K_WORK_DELAYABLE_BIT));
+	__ASSERT_NO_MSG(!k_is_in_isr());
+
+	struct z_work_canceller *canceller = &sync->canceller;
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	(void)cancel_async_locked(work);
+
+	bool need_wait = cancel_sync_locked(work, canceller);
+
+	k_spin_unlock(&lock, key);
+
+	if (need_wait) {
+		k_sem_take(&canceller->sem, K_FOREVER);
+	}
+
+	return need_wait;
+}
+
+/* Work has been dequeued and is about to be invoked by the work
+ * thread.
+ *
+ * If the work is being canceled the cancellation will be completed
+ * here, and the caller told not to use the work item.
+ *
+ * Invoked by work queue thread.
+ * Takes and releases lock.
+ * Reschedules via finalize_cancel_locked
+ *
+ * @param work work that is changing state
+ * @param queue queue that is running work
+ *
+ * @retval true if work is to be run by the work thread
+ * @retval false if it has been canceled and should not be run
+ */
+static inline bool work_set_running(struct k_work *work,
+				    struct k_work_q *queue)
+{
+	bool ret = false;
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	/* Allow the work to be queued again. */
+	atomic_clear_bit(&work->flags, K_WORK_QUEUED_BIT);
+
+	/* Normally we indicate that the work is being processed by
+	 * setting RUNNING.  However, something may have initiated
+	 * cancellation between when the work thread pulled this off
+	 * its queue and this claimed the work lock.  If that happened
+	 * we complete the cancellation now and tell the work thread
+	 * not to do anything.
+	 */
+	ret = !atomic_test_bit(&work->flags, K_WORK_CANCELING_BIT);
+	if (ret) {
+		/* Not cancelling: mark running and go */
+		atomic_set_bit(&work->flags, K_WORK_RUNNING_BIT);
+	} else {
+		/* Caught the item before being invoked; complete the
+		 * cancellation now.
+		 */
+		finalize_cancel_locked(work);
+	}
+
+	k_spin_unlock(&lock, key);
+
+	return ret;
+}
+
+/* Work handler has been called and is about to go idle.
+ *
+ * If the work is being canceled this will notify anything waiting
+ * for the cancellation.
+ *
+ * Invoked by work queue thread.
+ * Takes and releases lock.
+ * Reschedules via finalize_cancel_locked
+ *
+ * @param work work that is in running state
+ */
+static inline void work_clear_running(struct k_work *work)
+{
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	/* Clear running */
+	atomic_clear_bit(&work->flags, K_WORK_RUNNING_BIT);
+
+	if (atomic_test_bit(&work->flags, K_WORK_CANCELING_BIT)) {
+		finalize_cancel_locked(work);
+	}
+
+	k_spin_unlock(&lock, key);
+}
+
+/* Loop executed by a work queue thread.
+ *
+ * @param workq_ptr pointer to the work queue structure
+ */
+static void work_queue_main(void *workq_ptr, void *p2, void *p3)
+{
+	struct k_work_q *queue = (struct k_work_q *)workq_ptr;
+
+	while (true) {
+		sys_snode_t *node;
+		struct k_work *work = NULL;
+		k_spinlock_key_t key = k_spin_lock(&queue->lock);
```

Comment:
Actually the queue lock is taken in isolation in `k_work_queue_{drain,unplug}` as well.  Also it is documented in the definition of the lock field in `struct k_work_q`:

        /* Lock protecting pending pending, notifyq, drainq, and flags.
         *
         * This may be taken when the work lock is head.  No other locks may
         * be taken while this is held.
         */

But I've added a comment here anyway.

Reply:
And next update: s/head/held/